File: backend/src/services/auth.service.ts
import bcrypt from "bcryptjs";
import { prisma } from "../config/db";
import { ErrorCode } from "../enums";

export const registerUser = async (
  name: string,
  email: string,
  password: string,
) => {
  const existingUser = await prisma.user.findUnique({ where: { email } });

  if (existingUser) {
    throw { code: ErrorCode.USER_ALREADY_EXISTS };
  }

  const hashedPassword = await bcrypt.hash(password, 10);

  const user = await prisma.user.create({
    data: { name, email, password: hashedPassword },
  });

  return user;
};

export const loginUser = async (email: string, password: string) => {
  const user = await prisma.user.findUnique({ where: { email } });

  if (!user) {
    throw { code: ErrorCode.USER_NOT_FOUND };
  }

  const isValid = await bcrypt.compare(password, user.password);

  if (!isValid) {
    throw { code: ErrorCode.INVALID_CREDENTIALS };
  }

  return user;
};

File: backend/app/services/auth_service.py
from sqlalchemy.orm import Session
from fastapi import Response
from sqlalchemy.exc import IntegrityError

from app.models.user import User
from app.core.security import hash_password, verify_password, create_access_token
from app.core.errors import AppError, ErrorCode

from dotenv import load_dotenv
import os

from app.schemas.auth_schema import LoginRequest, RegisterRequest

load_dotenv()

ENV = os.getenv("ENV")


def register_user(db: Session, data: RegisterRequest):

    user = User(
        email=data.email,
        username=data.username,
        password=hash_password(data.password),
    )

    try:
        db.add(user)
        db.commit()
        db.refresh(user)
        return user
    except IntegrityError as e:
        db.rollback()

        error_str = str(e.orig).lower()

        if "email" in error_str:
            raise AppError(ErrorCode.EMAIL_ALREADY_EXISTS)

        if "username" in error_str:
            raise AppError(ErrorCode.USERNAME_ALREADY_EXISTS)

        raise AppError(ErrorCode.USER_ALREADY_EXISTS)


def login_user(db: Session, response: Response, data: LoginRequest):
    user = db.query(User).filter(User.email == data.email).first()

    if not user or not verify_password(data.password, user.password):
        raise AppError(ErrorCode.INVALID_CREDENTIALS)

    token = create_access_token({"sub": str(user.id)})

    response.set_cookie(
        key="access_token",
        value=token,
        httponly=True,
        secure=ENV == "prod",
        samesite="lax",
        max_age=60 * 60 * 24,  # 1 day
    )

    return user
